### PR TITLE
support non-strict mode to decode object to map when unregistered

### LINKED ISSUE #308: 希望hessian库提供一个接口，可以将[]byte 和 map 结构互转。
**What would you like to be added**:

希望hessian库提供一个接口，可以将[]byte 转为map实现，并可以将map恢复为[]byte，例如
```go
type User struct{
    Name string
    Age uint32
}

type (u *User) JavaClassName() string{
    return "org.apache.com.myUser"
}

enc := hessian.NewEncoder()
data1 := enc.Encode(&User{}) // encode to []byte

dec := hessian.NewDecoder(data) 

myMap := dec.DecodeMap() // todo []byte to map[string]interface{}
// myMap["name"] == "xxx"
// myMap["age"] == "11"

date2 := enc.EncodeMap(myMap, "org.apache.com.myUser") // data2 should be equal to data1
```

**Why is this needed**:

### PR DESCRIPTION
**Pull request by wongoo**
**What this PR does**:
- support non-strict mode to decode object to map when unregistered
- support encode object to class

**Which issue(s) this PR fixes**:

Fixes #308

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
- support non-strict mode to decode object to map when unregistered
```

example:
```go
       name := UserName{
		FirstName: "John",
		LastName:  "Doe",
	}
	person := Person{
		UserName: name,
		Age:      18,
		Sex:      true,
	}

	worker1 := &Worker{
		Person: person,
		CurJob: JOB{Title: "cto", Company: "facebook"},
		Jobs: []JOB{
			{Title: "manager", Company: "google"},
			{Title: "ceo", Company: "microsoft"},
		},
	}

	t.Logf("worker1: %v", worker1)

	e := NewEncoder()
	err := e.Encode(worker1)
	if err != nil {
		t.Fatalf("encode(worker:%#v) = error:%s", worker1, err)
	}
	data := e.Buffer()
	t.Logf("data: %s", data)

	// unRegisterPOJO before decode, so that to decode to map
	unRegisterPOJO(name)
	unRegisterPOJO(person)
	unRegisterPOJO(worker1)
	unRegisterPOJO(&worker1.Jobs[0])

	// strict mode
	d := NewDecoder(data)
	d.Strict = true // set to strict mode, then the decoding will return error for class not being registered.
	res, err := d.Decode()
	if err == nil {
		t.Error("after unregister pojo, decoding should return error for strict mode")
		t.FailNow()
	}

	// non-strict mode, decode to map for class not being registered.
	d = NewDecoder(data)
	res, err = d.Decode()
	if err != nil {
		t.Error(err)
		t.FailNow()
	}
	t.Logf("type of decode object:%v", reflect.TypeOf(res))

	worker2, ok := res.(map[string]interface{}) // the response is a map
	if !ok {
		t.Fatalf("res:%#v should be a map for non-strict mode", res)
	}

	t.Logf("worker2: %v", worker2)

	// register pojo again, so that there are class definitions, which are used to encode map to object.
	RegisterPOJO(name)
	RegisterPOJO(person)
	RegisterPOJO(worker1)
	RegisterPOJO(&worker1.Jobs[0])

	// encode the map to object again, note the worker2 is a map.
	e = NewEncoder()
	err = e.Encode(worker2)
	if err != nil {
		t.Error(err)
		t.FailNow()
	}

	data = e.Buffer()
	t.Logf("data: %s", data)

	// decode the encoded map data to struct object again.
	d = NewDecoder(data)
	res, err = d.Decode()
	if err != nil {
		t.Error(err)
		t.FailNow()
	}
	t.Logf("type of decode object:%v", reflect.TypeOf(res))

	worker3, ok := res.(*Worker) // worker3 is a Worker
	if !ok {
		t.Fatalf("res:%#v should be a worker type", res)
	}

	t.Logf("worker3: %v", worker3)
	if !reflect.DeepEqual(worker1, worker3) {
		t.Fatal("worker1 not equal to worker3!")
	}
```

### CODE REVIEW

File: response.go [hunk 1]
```diff
@@ -204,19 +199,15 @@ func unpackResponseBody(decoder *Decoder, resp interface{}) error {
 			}
 		}
 
-		// If the return value is nil,
-		// we should consider it normal
-		if rsp == nil {
-			return nil
-		}
+		response.RspObj = rsp
```
Comment (wongoo) [inline response.go:202]: @LaurenceLiZhixin @AlexStocks   just assign the decoded object to the response，I think this is the biggest change, which should be tested in dubbo-go.

**Review by zouyx — APPROVED** [review]

**Comment by LaurenceLiZhixin** [comment]
@wongoo 
I have a question, can we encode a map without any registry? I think the map[string]interface{} contains all information  about a pojo.
On the view of gateway filter, gateway can't get the real POJO of the object. But in the current implementation, this step is necessary.
```go
// register pojo again, so that there are class definitions, which are used to encode map to object.
/**
Can these be omited.
**/
RegisterPOJO(name)
RegisterPOJO(person)
RegisterPOJO(worker1)
RegisterPOJO(&worker1.Jobs[0])

// encode the map to object again, note the worker2 is a map.
e = NewEncoder()
err = e.Encode(worker2)
if err != nil {
t.Error(err)
t.FailNow()
}

```

File: response.go (hunk 1 above)
Comment (LaurenceLiZhixin) [reply response.go:202]: Got it.

**Comment by wongoo** [comment]
@LaurenceLiZhixin  POJO class definition is mandentary

**Comment by wongoo** [comment]
@LaurenceLiZhixin  yes, we can encode a map directly, I will try to implement it.

File: decode.go [hunk 2]
```diff
@@ -49,6 +64,16 @@ func NewDecoder(b []byte) *Decoder {
 	return &Decoder{reader: bufio.NewReader(bytes.NewReader(b)), typeRefs: &TypeRefs{records: map[string]bool{}}}
 }
 
+// NewStrictDecoder generate a strict mode decoder instance.
+// In strict mode, all target class must be registered.
```
Comment (AlexStocks) [inline decode.go:68]: generate -> generates

pls add this comment to readme to let the user know how to use it.

**Comment by wongoo** [comment]
@LaurenceLiZhixin  I implement build class definition from map keys, and add a unit test:

```go
func TestEncodeMapToObject(t *testing.T) {
	name := &UserName{
		FirstName: "John",
		LastName:  "Doe",
	}

	RegisterPOJO(name)

	// note: the first letter of the keys MUST lowercase.
	m := map[string]interface{}{
		"firstName": "John",
		"lastName":  "Doe",
	}

	e := NewEncoder()
	encErr := e.EncodeMapAsClass(name.JavaClassName(), m)
	if encErr != nil {
		t.Error(encErr)
		t.FailNow()
	}

	res := mustDecodeObject(t, e.Buffer())
	assert.True(t, reflect.DeepEqual(name, res))

	// note: the map contains the class key.
	m = map[string]interface{}{
		ClassKey:    name.JavaClassName(),
		"firstName": "John",
		"lastName":  "Doe",
	}

	// try to encode again
	e = NewEncoder()
	encErr = e.EncodeMapClass(m)
	if encErr != nil {
		t.Error(encErr)
		t.FailNow()
	}

	res = mustDecodeObject(t, e.Buffer())
	assert.True(t, reflect.DeepEqual(name, res))
}
```

**Review by LaurenceLiZhixin — APPROVED** [review]

**Comment by LaurenceLiZhixin** [comment]
I have verified this feature the locally, and it works fine!
As Apache DUBBO PMC, Wongoo senior is very rigorous and efficient in dealing with new-coming issue, which is worth learning by community members. On behalf of business workmates and aliRPC-Go's wide users, I would like to express my sincere thanks.
@wongoo

**Review by zhaoyunxing92 — APPROVED** [review]

**Pull request merged**